nni/compression/pytorch/speedup/compress_modules.py

```diff
@@ -3,196 +3,364 @@
+    'Hardsigmoid': lambda module, masks: no_replace(module, masks),
+    'LogSigmoid': lambda module, masks: no_replace(module, masks),
+    'PReLU': lambda module, masks: no_replace(module, masks),
+    'RReLU': lambda module, masks: no_replace(module, masks),
+    'SELU': lambda module, masks: no_replace(module, masks),
+    'CELU': lambda module, masks: no_replace(module, masks),
+    'GELU': lambda module, masks: no_replace(module, masks),
+    'Sigmoid': lambda module, masks: no_replace(module, masks),
+    'SiLU': lambda module, masks: no_replace(module, masks),
+    'Mish': lambda module, masks: no_replace(module, masks),
+    'Tanh': lambda module, masks: no_replace(module, masks),
+    'Softplus': lambda module, masks: no_replace(module, masks),
+    'Softshrink': lambda module, masks: no_replace(module, masks),
+    'Softmax': lambda module, masks: no_replace(module, masks),
+    'Tanhshrink': lambda module, masks: no_replace(module, masks),
+    'Dropout': lambda module, masks: no_replace(module, masks),
+    'Dropout2d': lambda module, masks: no_replace(module, masks),
+    'Dropout3d': lambda module, masks: no_replace(module, masks),
+    'Upsample': lambda module, masks: no_replace(module, masks),
+    'LayerNorm': lambda module, masks: replace_layernorm(module, masks),
+    'ConvTranspose2d': lambda module, masks: replace_convtranspose2d(module, masks)
 }
 
 
-def no_replace(module, mask):
+
+def convert_to_coarse_mask(t_mask, dim):
+    """
+    Convert the mask tensor to the coarse-grained mask tensor.
+    Parameters
+    ---------
+    t_mask: torch.Tensor
+        The tensor only have 1s and 0s, 0 indicates this value is masked
+        and 1 indicates the corresponding value is not masked.
+    dim: int
+        Try to reduce the mask tensor on this dimension.
+
+    Returns
+    -------
+    indexes: torch.Tensor
+        The indexes of the sparsity that can be structurally removed.
+    remained_indexes: torch.Tensor
+        The indexes of values that need to be remained.
+    """
+    assert isinstance(t_mask, torch.Tensor)
+    shape = list(t_mask.size())
+    n_dims = len(shape)
+    dim_list = list(range(n_dims))
+    # try to reduce the mask from the dim-th dimension
+    dim_list.remove(dim)
+
+    t_merged = torch.sum(t_mask, dim_list)
+    assert t_merged.size(0) == shape[dim]
+    all_pruned = t_merged == 0
+    need_remain = t_merged != 0
+    # return the indexes of the sparsity that can be removed
+    indexes = torch.nonzero(all_pruned, as_tuple=True)[0]
+    remained_indexes = torch.nonzero(need_remain, as_tuple=True)[0]
+    return indexes, remained_indexes
+
+
+def no_replace(module, masks):
     """
     No need to replace
     """
     _logger.debug("no need to replace")
     return module
 
 
-def replace_linear(linear, mask):
+def replace_linear(linear, masks):
     """
+    This function will replace the original linear according to
+    the infered masks. This function support the fine-grained and
```

Comment:
does this function support fine-grained sparsity?

Reply:
Yes, I update the `convert_to_coarse_mask` function, it will remove the whole column/row that happened be covered by the fine-grained sparsity.

Reply:
got it